### PR TITLE
fix(halo): fix missing comet prometheus metrics

### PR DESCRIPTION
**Pull request by corverroos**
Fix missing cometBFT prometheus metrics

issue:  #2207

### CODE REVIEW

File: halo/app/monitor.go [hunk 1]
```diff
@@ -202,30 +211,32 @@ func dirSize(path string) (int64, error) {
...
+) func(context.Context) error {
 	mux := http.NewServeMux()
 
-	// We align both with industry standards (endpoint /metrics)
-	// and with cometBFT (endpoint /).
-	mux.Handle("/metrics", promhttp.Handler())
-	mux.Handle("/", promhttp.Handler())
+	promHandler := promhttp.HandlerFor(
+		prometheus.DefaultGatherer,
+		promhttp.HandlerOpts{MaxRequestsInFlight: cfg.Instrumentation.MaxOpenConnections},
```
Comment (corverroos) [inline halo/app/monitor.go:226]: wire `cfg.Instrumentation.MaxOpenConnections` which was also not used anymore

File: halo/app/monitor.go [hunk 2]
```diff
@@ -202,30 +211,32 @@ func dirSize(path string) (int64, error) {
...
+	)
+	mux.Handle("/metrics", promHandler) // Industry standard `/metrics` path
+	mux.Handle("/", promHandler)        // CometBFT standard `/` path
 
-	// On the `/ready` endpoint, we serve the readiness of the halo node.
-	// Additionally, in case when the node is not ready, the response status code is set to 503.
+	// Serve readiness status json at `/ready`, returning 503 if not ready.
```
Comment (corverroos) [inline halo/app/monitor.go:231]: Hey, I adjusted the comment to make it more concise and direct. I also removed "we" to keep it neutral and focused on the action.

File: halo/app/monitor.go [hunk 3]
```diff
@@ -202,30 +211,32 @@ func dirSize(path string) (int64, error) {
...
 	mux.HandleFunc("/ready", func(w http.ResponseWriter, _ *http.Request) {
 		w.Header().Set("Content-Type", "application/json")
 
-		if !status.ready() {
-			w.WriteHeader(http.StatusServiceUnavailable)
-		}
-		if err := status.serialize(w); err != nil {
```
Comment (corverroos) [inline halo/app/monitor.go:226]: there is actualy a race condition here. We read ready and then we serialiaze. Technically, the status could change in between these two steps resulting in inconsistent output. This isn't critical code, but I thought I would clean it up in any case

File: halo/app/monitor.go [hunk 4]
```diff
@@ -236,8 +247,8 @@ func startMonitoringAPI(
 	}
 
 	go func() {
-		if err := server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
```
Comment (corverroos) [inline halo/app/monitor.go:239]: see `server.ListenAndServe` docs, it always returns a non-nil error, so no need to check for nil

File: halo/app/ready.go [hunk 5]
```diff
@@ -13,39 +13,40 @@ type readinessStatus struct {
 	mu                 sync.RWMutex
 	ConsensusSynced    bool    `json:"consensus_synced"`              // Unhealthy if false
 	ConsensusP2PPeers  int     `json:"consensus_p2p_peers"`           // Unhealthy if 0
+	ConsensusHeight    int64   `json:"consensus_height"`              // Unhealthy if 0
+	ConsensusRunning   bool    `json:"consensus_running"`             // Unhealthy if false
 	ExecutionConnected bool    `json:"execution_connected"`           // Unhealthy if false
 	ExecutionSynced    bool    `json:"execution_synced"`              // Unhealthy if false
-	ExecutionP2PPeers  *uint64 `json:"execution_p2p_peers,omitempty"` // Unhealthy if 0
+	ExecutionP2PPeers  *uint64 `json:"execution_p2p_peers,omitempty"` // Unhealthy if 0 (ignored if not available)
+	ExecutionHeight    uint64  `json:"execution_height"`              // Unhealthy if 0
```
Comment (corverroos) [inline halo/app/ready.go:21]: added Heights, just as a nice to have to users querying the endpoint and looking at the json.

File: halo/app/ready.go [hunk 6]
```diff
@@ -13,39 +13,40 @@ type readinessStatus struct {
 	mu                 sync.RWMutex
 	ConsensusSynced    bool    `json:"consensus_synced"`              // Unhealthy if false
 	ConsensusP2PPeers  int     `json:"consensus_p2p_peers"`           // Unhealthy if 0
+	ConsensusHeight    int64   `json:"consensus_height"`              // Unhealthy if 0
+	ConsensusRunning   bool    `json:"consensus_running"`             // Unhealthy if false
```
Comment (corverroos) [inline halo/app/ready.go:17]: aded `consensus_running` so the app becomes unhealthy if consensus stalls like on staging. The nodes were healthy even though they stalled

File: halo/app/start.go [hunk 7]
```diff
@@ -314,7 +314,12 @@ func newCometNode(ctx context.Context, cfg *cmtcfg.Config, app *App, privVal cmt
 		},
 	)
 
-	// Don't instantiate the default prometheus server.
+	// Configure CometBFT prometheus metrics as per provided config
+	metrics := node.DefaultMetricsProvider(&cmtcfg.InstrumentationConfig{
+		Prometheus: cfg.Instrumentation.Prometheus,
+		Namespace:  cfg.Instrumentation.Namespace,
```
Comment (corverroos) [inline halo/app/start.go:320]: I think is that you were missing in your PR. That resulted in incorrect metric names so it seemed like it wasn't renabled.

Comment (chmllr) [reply halo/app/start.go:320]: Yes that must be it because I saw there is no prefix `cometbft_` 👍

Thanks a lot Corver!

File: halo/app/ready.go [hunk 8]
```diff
@@ -13,39 +13,40 @@ type readinessStatus struct {
...
 	r.mu.RLock()
 	defer r.mu.RUnlock()
 
 	if err := json.NewEncoder(w).Encode(r); err != nil {
-		return errors.Wrap(err, "readiness status serialization")
+		return false, errors.Wrap(err, "readiness status serialization")
 	}
 
-	return nil
+	return r.ready(), nil
```
Comment (chmllr) [inline halo/app/ready.go:49]: Wouldn't this cause a dead lock? `ready()` will try to lock while we're holding the lock?

Comment (chmllr) [reply halo/app/ready.go:49]: Ah looks like read locks are reentrant in Go by default. This explains why no test is failing.

**Review by chmllr — APPROVED** [review]

**Pull request merged**